File: slim3/src/main/java/jp_zipcode_search/model/Zip.java
package jp_zipcode_search.model;

import com.google.appengine.api.datastore.Key;
import org.slim3.datastore.Attribute;
import org.slim3.datastore.Model;

@Model
public class Zip {

  @Attribute(primaryKey = true)
  private Key key;

  private String zipcode;

  private String pref;

  private String city;

  private String address;

  private String pref_phonetic;

  private String city_phonetic;

  private String address_phonetic;

  public Key getKey() {
    return key;
  }

  public void setKey(Key key) {
    this.key = key;
  }

  public String getZipcode() {
    return zipcode;
  }

  public void setZipcode(String zipcode) {
    this.zipcode = zipcode;
  }

  public String getPref() {
    return pref;
  }

  public void setPref(String pref) {
    this.pref = pref;
  }

  public String getCity() {
    return city;
  }

  public void setCity(String city) {
    this.city = city;
  }

  public String getAddress() {
    return address;
  }

  public void setAddress(String address) {
    this.address = address;
  }

  public String getPref_phonetic() {
    return pref_phonetic;
  }

  public void setPref_phonetic(String pref_phonetic) {
    this.pref_phonetic = pref_phonetic;
  }

  public String getCity_phonetic() {
    return city_phonetic;
  }

  public void setCity_phonetic(String city_phonetic) {
    this.city_phonetic = city_phonetic;
  }

  public String getAddress_phonetic() {
    return address_phonetic;
  }

  public void setAddress_phonetic(String address_phonetic) {
    this.address_phonetic = address_phonetic;
  }

}

File: lift/src/main/scala/jp_zipcode_search/model/Zip.scala
package jp_zipcode_search.model

import net.liftweb.util._
import java.util._
import scala.collection.jcl.Conversions._
import com.google.appengine.api.datastore._
import Query._
import FilterOperator._
import SortDirection._
import FetchOptions.Builder._

class Zip(entity:Entity) {
  
  def this() = this(new Entity("Zip"))
  
  def zipcode = entity.getProperty("zipcode").toString
  
  def prefPhonetic = entity.getProperty("pref_phonetic").toString
  
  def cityPhonetic = entity.getProperty("city_phonetic").toString
  
  def addressPhonetic = entity.getProperty("address_phonetic").toString
  
  def pref = entity.getProperty("pref").toString
  
  def city = entity.getProperty("city").toString
  
  def address = entity.getProperty("address").toString

}

object Zip {
  
  val ds:DatastoreService = DatastoreServiceFactory.getDatastoreService
  
  def create(entity:Entity) = new Zip(entity)
  
  def find(address:scala.List[String]) = {
    var query = new Query("Zip")
    if (address.length > 0) {
      query.addFilter("pref", EQUAL, address(0))
    }
    if (address.length > 1) {
      query.addFilter("city", EQUAL, address(1))
    }
    if (address.length > 2) {
      query.addFilter("address", GREATER_THAN_OR_EQUAL, address(2))
      query.addFilter("address", LESS_THAN, address(2) + "\ufffd")
      query.addSort("address", ASCENDING)
    }
    query.addSort("city_phonetic", ASCENDING)
    query.addSort("address_phonetic", ASCENDING)
    ds.prepare(query).asList(withOffset(0).limit(1000)).toList.map(create)
  }
  
  def findByZipcode(zipcode:String) = {
    var query = new Query("Zip")
    query.addSort("city_phonetic", ASCENDING)
    query.addSort("address_phonetic", ASCENDING)
    query.addFilter("zipcode", EQUAL, zipcode)
    ds.prepare(query).asList(withOffset(0)).toList.map(create)
  }
  
}